### PR TITLE
[ML] fix array oob in IDGenerator and adjust format for mapping

### LINKED ISSUE #41694: [ML] Data frame transform fails to handle timestamp formatting
Indexing documents fails in the following situation:
 - When using a `date_histogram` in the `group_by` definition
 - When a `format` is supplied in that definition
    - Example `"format": "yyyy-MM-dd HH:mm"`
 - When the source field is mapped as a `date` type

When we determine the destination mapping, we correct set it to `date`. Regretfully, we do not provide the appropriate `format` field in the mapping definition to account for the DateHistogram bucket keys having the format as described. 

We should adjust `SchemaUtil#deduceMappings` or `DataframeIndex#createDestinationIndex` to account for different possible formats for timestamp fields.

### PR DESCRIPTION
**Pull request by benwtrent**
There are two bugs addressed here:

* Sometimes when generating an ID, the resulting Byte array could be empty (for example, this occurs with an `""` empty string). We should handle that scenario by not allowing the byte array to be empty
* When the user provides a non-standard format for the `date_histogram` group by pivot, we do not consider it when we create the mapping. We should add their provided format as one of the possible formats for the resulting `date` field. 

closes https://github.com/elastic/elasticsearch/issues/41694

### CODE REVIEW

File: x-pack/plugin/data-frame/qa/single-node-tests/src/test/java/org/elasticsearch/xpack/dataframe/integration/DataFramePivotRestIT.java [hunk 1]
```diff
@@ -234,8 +234,8 @@ public void testDateHistogramPivot() throws Exception {
 
         // we expect 21 documents as there shall be 21 days worth of docs
```
Comment (davidkyle) [inline x-pack/plugin/data-frame/qa/single-node-tests/src/test/java/org/elasticsearch/xpack/dataframe/integration/DataFramePivotRestIT.java]: nit remove or update the comment

**Review by davidkyle — APPROVED** [review]
LGTM

**Review by davidkyle — APPROVED** [review]

File: x-pack/plugin/data-frame/qa/single-node-tests/src/test/java/org/elasticsearch/xpack/dataframe/integration/DataFramePivotRestIT.java (hunk 1 above)
Comment (benwtrent) [reply x-pack/plugin/data-frame/qa/single-node-tests/src/test/java/org/elasticsearch/xpack/dataframe/integration/DataFramePivotRestIT.java]: lulz. Maybe I should just stop writing comments :D

**Comment by benwtrent** [comment]
run elasticsearch-ci/bwc

**Pull request merged**